feat: unified swaps

infiniteflower opened the pull request:
<!--
Please submit this PR as a draft initially.
Do not mark it as "Ready for review" until the template has been completely filled out, and PR status checks have passed at least once.
-->

## **Description**

<!--
Write a short description of the changes included in this pull request, also include relevant motivation and context. Have in mind the following questions:
1. What is the reason for the change?
2. What is the improvement/solution?
-->

This PR handles Unified Swaps routing, network, and token selection. It's hidden behind an env var and a feature flag, so there is no risk of accidentally exposing this feature before it's ready.

## **Related issues**

Resolves: https://consensyssoftware.atlassian.net/browse/MMS-2341

## **Manual testing steps**

Setup

1. Add `export MM_UNIFIED_SWAPS_ENABLED="true"` to `.js.env`
2. Terminal: `source .js.env`
3. Start the app up

Swaps

5. Go to Swaps (notice there's no Bridge button anymore)
6. Request a quote for a source and dest token with the same chain

Bridge

5. Go to Swaps (notice there's no Bridge button anymore)
6. Request a quote for a source and dest token with a different chain

Feature flags

1. Add `export MM_UNIFIED_SWAPS_ENABLED="false"` to `.js.env`
2. Terminal: `source .js.env`
3. Start the app up
4. You should see 2 buttons for Swap and Bridge
5. Swap btn should take you to Legacy Swaps
6. Bridge btn should take you to Bridge
 
## **Screenshots/Recordings**

<!-- If applicable, add screenshots and/or recordings to visualize the before and after of your change. -->

### **Before**

<!-- [screenshots/recordings] -->

### **After**

<!-- [screenshots/recordings] -->

Bridge

https://github.com/user-attachments/assets/eb597e89-8e98-4413-a006-f17a5428e1c2

Swap

https://github.com/user-attachments/assets/1c7dd607-7653-4359-bd5a-53d528ba5b5c

## **Pre-merge author checklist**

- [x] I’ve followed [MetaMask Contributor Docs](https://github.com/MetaMask/contributor-docs) and [MetaMask Mobile Coding Standards](https://github.com/MetaMask/metamask-mobile/blob/main/.github/guidelines/CODING_GUIDELINES.md).
- [x] I've completed the PR template to the best of my ability
- [x] I’ve included tests if applicable
- [x] I’ve documented my code using [JSDoc](https://jsdoc.app/) format if applicable
- [x] I’ve applied the right labels on the PR (see [labeling guidelines](https://github.com/MetaMask/metamask-mobile/blob/main/.github/guidelines/LABELING_GUIDELINES.md)). Not required for external contributors.

## **Pre-merge reviewer checklist**

- [ ] I've manually tested the PR (e.g. pull and build branch, run the app, test code being changed).
- [ ] I confirm that this PR addresses all acceptance criteria described in the ticket it closes and includes the necessary testing evidence such as recordings and or screenshots.

infiniteflower committed:
chore: route to bridge unified view on swaps press if unified enabled

infiniteflower committed:
chore: set header correctly for unified

infiniteflower committed:
chore: show all networks including current for dest network selector

infiniteflower committed:
fix: source networks bar not appearing in unified

infiniteflower committed:
chore: show current network in dest network bar

infiniteflower committed:
chore: simplify bridgeViewMode props passing and move it to redux instead

infiniteflower committed:
chore: make code more readable

infiniteflower committed:
fix: incorrect confirm button text

infiniteflower committed:
feat: consume LD flags for unified UI

infiniteflower committed:
fix: broken tests

infiniteflower committed:
fix: broken tests

infiniteflower committed:
fix: broken tests

infiniteflower committed:
fix: broken tests

infiniteflower committed:
fix: lint errors

infiniteflower committed:
fix: broken tests

infiniteflower committed:
chore: add tests for unified routing

infiniteflower committed:
chore: add tests for bridgeViewMode redux slice

infiniteflower committed:
fix: broken tests

infiniteflower committed:
chore: if unified enabled, only show swap button

infiniteflower committed:
feat: properly display unified swaps tx history for native gas token

infiniteflower committed:
fix: swapping from assets screen for erc20s not on mainnet crashing

infiniteflower committed:
fix: source amount sent being a massive number

infiniteflower committed:
fix: broken tests

infiniteflower committed:
chore: clean up code

infiniteflower committed:
chore: add test

infiniteflower committed:
fix: improper renderValue

infiniteflower committed:
chore: remove unneded test lines

infiniteflower committed:
chore: remove more lines

infiniteflower committed:
fix: bridge decoding not working properly

infiniteflower committed:
fix: improper tx type, and add tests

infiniteflower committed:
fix: broken tests

bfullam reviewed (changes requested):
LGTM. Only issue I see is that we're still using "Bridge to" as the button text for when the dest token is empty even though all other text uses "Swap". Should probably be changed to "Swap to", no?

infiniteflower committed:
fix: unified swaps button should say swap to

infiniteflower committed:
fix: broken tests

infiniteflower committed:
feat: use bridge tx details screen for unified swaps details as well

infiniteflower committed:
chore: remove unneeded test data

infiniteflower committed:
chore: remove more lines

infiniteflower committed:
chore: remove more lines

infiniteflower committed:
chore: send user directly to block explorer for unified swaps, no modal

infiniteflower committed:
chore: add tests for unified swaps selector

infiniteflower committed:
fix: issue where bridge to network wouldnt show up properly

infiniteflower committed:
fix: swaps appearing as a bridge (#16548)

<!--
Please submit this PR as a draft initially.
Do not mark it as "Ready for review" until the template has been
completely filled out, and PR status checks have passed at least once.
-->

## **Description**

<!--
Write a short description of the changes included in this pull request,
also include relevant motivation and context. Have in mind the following
questions:
1. What is the reason for the change?
2. What is the improvement/solution?
-->

This PR bumps BridgeController to 33.0.1 and BridgeStatusController to
32.0.0 in order to fix issues with Swaps on non Ethereum mainnet
appearing as Bridges. It also bumps all the required peer deps and their
deps.

## **Related issues**

Related to https://github.com/MetaMask/metamask-mobile/pull/16433

## **Manual testing steps**

Setup

1. Add `export MM_UNIFIED_SWAPS_ENABLED="true"` to `.js.env`
2. Terminal: `source .js.env`
3. Start the app up

1. Go Swaps
2. Select a token from not ethereum mainnet
3. Swap to a different token on the same chain
4. Go to tx history
5. See that the swap should be fine

## **Screenshots/Recordings**

<!-- If applicable, add screenshots and/or recordings to visualize the
before and after of your change. -->

### **Before**

<!-- [screenshots/recordings] -->

### **After**

<!-- [screenshots/recordings] -->
<img width="617" alt="Screenshot 2025-06-20 at 3 01 35 PM"
src="https://github.com/user-attachments/assets/f8d1d140-d7ba-46d4-a5bb-8a37592e6731"
/>



## **Pre-merge author checklist**

- [x] I’ve followed [MetaMask Contributor
Docs](https://github.com/MetaMask/contributor-docs) and [MetaMask Mobile
Coding
Standards](https://github.com/MetaMask/metamask-mobile/blob/main/.github/guidelines/CODING_GUIDELINES.md).
- [x] I've completed the PR template to the best of my ability
- [ ] I’ve included tests if applicable
- [ ] I’ve documented my code using [JSDoc](https://jsdoc.app/) format
if applicable
- [x] I’ve applied the right labels on the PR (see [labeling
guidelines](https://github.com/MetaMask/metamask-mobile/blob/main/.github/guidelines/LABELING_GUIDELINES.md)).
Not required for external contributors.

## **Pre-merge reviewer checklist**

- [ ] I've manually tested the PR (e.g. pull and build branch, run the
app, test code being changed).
- [ ] I confirm that this PR addresses all acceptance criteria described
in the ticket it closes and includes the necessary testing evidence such
as recordings and or screenshots.

infiniteflower committed:
fix: solana swaps not going to bridge tx details

infiniteflower committed:
Merge branch 'main' into mms-2341-unified-swaps

infiniteflower committed:
chore: bump yarn lock

amitabh94 commented:
bugbot run

infiniteflower committed:
fix: possible race condition with navigation and bridgeViewMode state

infiniteflower committed:
fix: not being able to enter unified swaps in e2e

infiniteflower committed:
fix: change txController back to 57.4.0 to fix bridge e2e

infiniteflower committed:
chore: update bitrise.yml with unified env vars

infiniteflower committed:
chore: skip legacy swaps e2e

infiniteflower committed:
fix: more bridge e2e

infiniteflower committed:
fix: tx submission e2e tests

infiniteflower committed:
chore: update snapshot

infiniteflower committed:
chore: skip metrics test

infiniteflower committed:
chore: update snapshot

salimtb reviewed (dismissed):
Assets changes LGTM

bfullam committed:
Merge branch 'main' into mms-2341-unified-swaps

tommasini commented on bitrise.yml:
```diff
@@ -2464,10 +2464,13 @@ app:
     - opts:
         is_expand: false
-      BRIDGE_USE_DEV_APIS: false
+      BRIDGE_USE_DEV_APIS: true
```
By the PR description I understand that this should be false by default

infiniteflower replied:
This one should be `true` this only really means we hit the dev Bridge API instead of the prod one

tommasini commented on bitrise.yml:
```diff
@@ -2464,10 +2464,13 @@ app:
       MM_BRIDGE_ENABLED: true
+    - opts:
+        is_expand: false
+      MM_UNIFIED_SWAPS_ENABLED: true
```
By the PR description I understand that this sould be false by default

infiniteflower committed:
chore: bump snaps to 14

infiniteflower commented on package.json:
```diff
@@ -173,11 +173,11 @@
     "@metamask/app-metadata-controller": "^1.0.0",
     "@metamask/approval-controller": "^7.1.3",
-    "@metamask/assets-controllers": "^68.2.0",
+    "@metamask/assets-controllers": "^69.0.0",
```
Bumped Snaps to 14

infiniteflower replied:
Actually Snaps team said I can stay on Snaps 13, they will handle the 14 bump in this PR https://github.com/MetaMask/metamask-mobile/pull/16911

infiniteflower committed:
Merge branch 'main' into mms-2341-unified-swaps

infiniteflower committed:
chore: yarn dedupe

davibroc committed:
Refactored the e2e tests (#16611)

<!--
Please submit this PR as a draft initially.
Do not mark it as "Ready for review" until the template has been
completely filled out, and PR status checks have passed at least once.
-->

## **Description**

<!--
Write a short description of the changes included in this pull request,
also include relevant motivation and context. Have in mind the following
questions:
1. What is the reason for the change?
2. What is the improvement/solution?
-->

## **Related issues**

Fixes:

## **Manual testing steps**

1. Go to this page...
2.
3.

## **Screenshots/Recordings**

<!-- If applicable, add screenshots and/or recordings to visualize the
before and after of your change. -->

### **Before**

<!-- [screenshots/recordings] -->

### **After**

<!-- [screenshots/recordings] -->

## **Pre-merge author checklist**

- [ ] I’ve followed [MetaMask Contributor
Docs](https://github.com/MetaMask/contributor-docs) and [MetaMask Mobile
Coding
Standards](https://github.com/MetaMask/metamask-mobile/blob/main/.github/guidelines/CODING_GUIDELINES.md).
- [x] I've completed the PR template to the best of my ability
- [ ] I’ve included tests if applicable
- [ ] I’ve documented my code using [JSDoc](https://jsdoc.app/) format
if applicable
- [ ] I’ve applied the right labels on the PR (see [labeling
guidelines](https://github.com/MetaMask/metamask-mobile/blob/main/.github/guidelines/LABELING_GUIDELINES.md)).
Not required for external contributors.

## **Pre-merge reviewer checklist**

- [x] I've manually tested the PR (e.g. pull and build branch, run the
app, test code being changed).
- [x] I confirm that this PR addresses all acceptance criteria described
in the ticket it closes and includes the necessary testing evidence such
as recordings and or screenshots.

---------

Co-authored-by: IF <139582705+infiniteflower@users.noreply.github.com>

infiniteflower committed:
chore: bump to tx 58

infiniteflower committed:
Merge branch 'main' into mms-2341-unified-swaps

infiniteflower committed:
chore: bump snaps to 14.0.2

infiniteflower committed:
chore: go back to snaps 13

darkwing reviewed (dismissed):
Approving Assets files.

OGPoyraz committed:
Merge branch 'main' into mms-2341-unified-swaps

bfullam committed:
Merge branch 'main' into mms-2341-unified-swaps

infiniteflower committed:
fix: attempt at fixing network abstraction E2E

bfullam committed:
Merge branch 'main' into mms-2341-unified-swaps

infiniteflower committed:
fix: network abstraction e2e fixes

infiniteflower committed:
chore: skip swap e2e

infiniteflower committed:
fix: lint error

bfullam committed:
Merge branch 'main' into mms-2341-unified-swaps

bfullam committed:
Merge branch 'main' into mms-2341-unified-swaps

bfullam committed:
Merge branch 'main' into mms-2341-unified-swaps

bfullam committed:
Merge branch 'main' into mms-2341-unified-swaps

bfullam committed:
fix: remove deprecated bridge status controller action stopPollingForQuotes

infiniteflower committed:
Merge branch 'main' into mms-2341-unified-swaps

infiniteflower committed:
fix: versions of bridge and bridge status got bumped down on a main merge, set to proper versions

infiniteflower committed:
chore: add random wallet to e2e setup

davibroc committed:
Mocked APIs for swap e2e tests

infiniteflower committed:
chore: remove eth > weth for now

infiniteflower committed:
fix: e2e

infiniteflower committed:
fix: e2e

tommasini reviewed (approved):
Left a comment, but if I'm missing something and we do not need to address it, feel free to resolve it

bfullam commented on app/components/Views/WalletActions/WalletActions.tsx:
```diff
@@ -413,7 +415,7 @@ const WalletActions = () => {
           />
         )}
-        {AppConstants.BRIDGE.ACTIVE && isBridgeAllowed(chainId) && (
+        {AppConstants.BRIDGE.ACTIVE && isBridgeAllowed(chainId) && !isUnifiedSwapsEnabled && (
```
Its the opposite, actually! If unified is enabled, we **do not** show the bridge action because you can do both swap and bridge within the unified UI. If unified is **disabled**, then we have the option to show both "swap" and "bridge" buttons separately 🤌
